Some test events sometimes not passed to sbt Test Reporter

When using ZIO Test, some test events are sometimes not passed to sbt's test reporter. I observed this behavior in sbt 1.9.6.

# Reproduction steps:
Copy these files, or clone https://github.com/Duhemm/zio-test-events-repro.

`build.sbt`:
```scala
lazy val checkReport = taskKey[Unit]("Check the test report")

lazy val repro = project.in(file("repro")).settings(
  libraryDependencies += "dev.zio" %% "zio-test" % "2.0.18" % Test,
  libraryDependencies += "dev.zio" %% "zio-test-sbt" % "2.0.18" % Test,
  testFrameworks += new TestFramework("zio.test.sbt.ZTestFramework"),

  checkReport := {
      val _ = (Test / test).value
      val reportsDirectory = (Test / testReportsDirectory).value
      val report = reportsDirectory / "TEST-com.example.ReproSpec.xml"
      val document = scala.xml.parsing.ConstructingParser.fromFile(report, false).document
      val cases = document \ "testcase"
      val length = cases.length
      assert(
        length == 2,
        s"Expected 2 test cases, found $length in the report. Report: $report"
      )
  }
)
```

`repro/src/test/scala/com/example/ReproSpec.scala`:
```scala
package com.example

import zio.test.{ZIOSpecDefault, _}
object ReproSpec extends ZIOSpecDefault{
    def spec = suite("int and string")(
        suite("string suite")(
            test("concat")(assertTrue("a" + "b" == "ab")),
            test("length")(assertTrue("abc".length == 3))
        )
    )
}
```

Run `checkReport` several times in sbt.

# Expected behavior
The `checkReport` task consistently succeeds because the test report consistently contains the 2 test cases.

# Actual behavior
Eventually, the `checkReport` fails because the report is missing one test case.

# Example output:
```
sbt:ziotest-missing-events> checkReport
+ int and string
  + string suite
    + concat
    + length
2 tests passed. 0 tests failed. 0 tests ignored.

Executed in 74 ms

[info] Completed tests
[success] Total time: 0 s, completed 6 Oct 2023, 16:39:26
sbt:ziotest-missing-events> checkReport
+ int and string
  + string suite
    + concat
    + length
2 tests passed. 0 tests failed. 0 tests ignored.

Executed in 74 ms

[info] Completed tests
[error] stack trace is suppressed; run last repro / checkReport for the full output
[error] (repro / checkReport) java.lang.AssertionError: assertion failed: Expected 2 test cases, found 1 in the report. Report: /Users/martin/Desktop/ziotest-missing-events/repro/target/test-reports/TEST-com.example.ReproSpec.xml
[error] Total time: 0 s, completed 6 Oct 2023, 16:39:26
sbt:ziotest-missing-events> checkReport
+ int and string
  + string suite
    + concat
    + length
2 tests passed. 0 tests failed. 0 tests ignored.

Executed in 74 ms

[info] Completed tests
[success] Total time: 0 s, completed 6 Oct 2023, 16:39:27
```

--- a/test-sbt/shared/src/main/scala/zio/test/sbt/ZTestEventHandlerSbt.scala
+++ b/test-sbt/shared/src/main/scala/zio/test/sbt/ZTestEventHandlerSbt.scala
@@ -2,7 +2,7 @@ package zio.test.sbt
 
 import sbt.testing.{EventHandler, Status, TaskDef}
 import zio.test.render.TestRenderer
-import zio.{UIO, ZIO}
+import zio.{Semaphore, UIO, Unsafe, ZIO}
 import zio.test.{ExecutionEvent, TestAnnotation, TestFailure, ZTestEventHandler}
 
 /**
@@ -16,13 +16,15 @@ import zio.test.{ExecutionEvent, TestAnnotation, TestFailure, ZTestEventHandler}
  */
 class ZTestEventHandlerSbt(eventHandler: EventHandler, taskDef: TaskDef, renderer: TestRenderer)
     extends ZTestEventHandler {
+  val semaphore = Semaphore.unsafe.make(1L)(Unsafe.unsafe)
   def handle(event: ExecutionEvent): UIO[Unit] =
     event match {
       // TODO Is there a non-sbt version of this I need to add similar handling to?
       case evt @ ExecutionEvent.TestStarted(_, _, _, _, _) =>
         ZIO.unit
       case evt @ ExecutionEvent.Test(_, _, _, _, _, _, _) =>
-        ZIO.succeed(eventHandler.handle(ZTestEvent.convertEvent(evt, taskDef, renderer)))
+        val zTestEvent = ZTestEvent.convertEvent(evt, taskDef, renderer)
+        semaphore.withPermit(ZIO.succeed(eventHandler.handle(zTestEvent)))
       case ExecutionEvent.SectionStart(_, _, _) => ZIO.unit
       case ExecutionEvent.SectionEnd(_, _, _)   => ZIO.unit
       case ExecutionEvent.TopLevelFlush(_)      => ZIO.unit
@@ -38,7 +40,7 @@ class ZTestEventHandlerSbt(eventHandler: EventHandler, taskDef: TaskDef, rendere
               annotations.get(TestAnnotation.timing).toMillis,
               ZioSpecFingerprint
             )
-            ZIO.succeed(eventHandler.handle(zTestEvent))
+            semaphore.withPermit(ZIO.succeed(eventHandler.handle(zTestEvent)))
         }
 
     }

